### PR TITLE
port of fixes related to #2112 from release 5.0.x

### LINKED ISSUE #2112: QueryDatabaseAuthenticationHandler not using passwordEncoder
- [5.0.0] CAS server version
- Embedded Tomcat on OSx
- Configure cas.authn.jdbc.query[0].passwordEncoder.type=BCRYPT and use DatabaseAuthentication query. Authentication is not successful due code in **QueryDatabaseAuthenticationHandler** class

` if(!StringUtils.equals(password, e))` because that is not right way to check if there is match. Instead passwordEncoder.matches should be used where first argument is raw password as provided by user. Reason is that BCryptPasswordEncoder will not encode same password to same string so comparing them is irrelevant.

### PR DESCRIPTION
**Pull request by rohajda**
Closes #2112 IssueNumber
port fixes from release 5.0.x

### CODE REVIEW

**Comment by mmoayyed** [comment]
Yup these are not related. You might want to take a pass at some of the comments posted here and fix those. I'll review the test failures.

File: core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import org.apereo.cas.authentication.UsernamePasswordCredential;
 import org.apereo.cas.authentication.handler.PrincipalNameTransformer;
 import org.apereo.cas.authentication.support.PasswordPolicyConfiguration;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
```
Comment (rohajda) [inline core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java]: fixed

File: core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java [hunk 2]
```diff
@@ -60,7 +61,25 @@ protected HandlerResult doAuthentication(final Credential credential) throws Gen
...
+     * @param transformedCredential the credential object bearing the transformed username and password.
+     * @param originalPassword original password from credential before password encoding
+     *
+     * @return HandlerResult resolved from credential on authentication success or null if no principal could be resolved
+     * from the credential.
+     *
+     * @throws GeneralSecurityException On authentication failure.
+     * @throws PreventedException On the indeterminate case when authentication is prevented.
+     */
+    protected HandlerResult authenticateUsernamePasswordInternal(UsernamePasswordCredential transformedCredential,final String originalPassword)
```
Comment (rohajda) [inline core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java]: fixed

File: support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java [hunk 3]
```diff
@@ -109,7 +112,7 @@ public void verifyMultipleRecords() throws Exception {
         q.setDataSource(this.dataSource);
         q.setSql(SQL);
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"), "psw0");
```
Comment (rohajda) [inline support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java]: corrected for new tests verifyBCryptSucces and verifyBCryptFail

File: support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java [hunk 4]
```diff
@@ -131,4 +134,42 @@ public void verifySuccess() throws Exception {
...
+        q.setPasswordEncoder(encoder);
+        q.authenticateUsernamePasswordInternal(
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+    }
+
+    /**
+     *  This test proves that in case BCRYPT and using raw password test can authenticate
+     */
+    @Test
+    public void verifyBCryptSuccess() throws Exception {
```
Comment (rohajda) [inline support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java]: fixed

**Comment by mmoayyed** [comment]
Thanks very much!

File: core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java (hunk 1 above)
Comment (rohajda) [inline core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java]: fixed

File: support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java (hunk 3 above)
Comment (rohajda) [inline support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java]: fixed

**Pull request merged**